MarkBind/markbind
# Add multiple test site support

**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [X] New feature



Fixes #574 



**What is the rationale for this request?**

Some features, such as tagging, must be tested on different sites since the test cases are mutually exclusive.

**What changes did you make? (Give an overview)**

Allow support for any number of test sites.

Revamped test directory structure:

- Old:
```
test/
  unit/
    (unit tests)
  test_site/
    (site files)
    test.sh
    test.bat
    testUtil/
      (test utils)
```

- New:
```
test/
  unit/
    (unit tests)
  functional/
    test.sh
    test.bat
    testUtil/
      (test utils)
    test_site/
      (test site files)
    test_site_2/
      (test site 2 files)
    test_site etc... /
      (...)
```

Creating a new test site:

- Create a folder in `test/functional`
- Run `markbind init` in the folder, write test site as per normal
- Create an `expected` directory within the folder and populate with expected files
- Add the folder name to these two lines:

test.sh
```
set sites=test_site test_site_2   // Add the folder name here
```

test.bat
```
declare -a sites=("test_site" "test_site_2")   // Add the folder name here
```

Note: Ensure `expected` directory is ignored (Add `expected/*` to the site.json ignore), and that no `src` or `glob` matches any of the expected files (the default glob `**/*.+(md|mbd)` causes some problems)

**Is there anything you'd like reviewers to focus on?**

The current `test_site_2` folder is just there to show the feature works, it stores a fresh init site. Will remove before merge. No edits made to test_site, except for removing 3 lines from the site.json

How do we want to organize the tests? For example, say we have:
- Feature A: Test case A1 and Test case A2 are mutually exclusive
- Feature B: Test case B1 and Test case B2 are mutually exclusive

We could do something like this:
- Site 1: Other tests
- Site 2: A1, B1
- Site 3: A2, B2
